src/QirRuntime/lib/QIR/bridge-qis.ll

```diff
@@ -29,122 +29,264 @@
-declare void @quantum__qis__x(%class.QUBIT*)
-declare void @quantum__qis__y(%class.QUBIT*)
-declare void @quantum__qis__z(%class.QUBIT*)
-
-declare void @quantum__qis__crx(%struct.QirArray*, double, %class.QUBIT*)
-declare void @quantum__qis__crz(%struct.QirArray*, double, %class.QUBIT*)
+
+declare void @quantum__qis__exp__body(%struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__exp__adj(%struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__exp__ctl(%struct.QirArray*, %struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__exp__ctladj(%struct.QirArray*, %struct.QirArray*, double, %struct.QirArray*)
+declare void @quantum__qis__h__body(%class.QUBIT*)
+declare void @quantum__qis__h__ctl(%struct.QirArray*, %class.QUBIT*)
+declare %class.RESULT* @quantum__qis__measure__body(%struct.QirArray*, %struct.QirArray*)
+declare void @quantum__qis__r__body(i32, double, %class.QUBIT*)
+declare void @quantum__qis__r__adj(i32, double, %class.QUBIT*)
+declare void @quantum__qis__r__ctl(%struct.QirArray*, i32, double, %class.QUBIT*)
+declare void @quantum__qis__r__ctladj(%struct.QirArray*, i32, double, %class.QUBIT*)
+declare void @quantum__qis__s__body(%class.QUBIT*)
+declare void @quantum__qis__s__adj(%class.QUBIT*)
+declare void @quantum__qis__s__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__s__ctladj(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__t__body(%class.QUBIT*)
+declare void @quantum__qis__t__adj(%class.QUBIT*)
+declare void @quantum__qis__t__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__t__ctladj(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__x__body(%class.QUBIT*)
+declare void @quantum__qis__x__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__y__body(%class.QUBIT*)
+declare void @quantum__qis__y__ctl(%struct.QirArray*, %class.QUBIT*)
+declare void @quantum__qis__z__body(%class.QUBIT*)
+declare void @quantum__qis__z__ctl(%struct.QirArray*, %class.QUBIT*)
 
 ;===============================================================================
 ; quantum.qis namespace implementations
 ;
-define double @__quantum__qis__intAsDouble(i64 %i)
-{
-  %d = call double @quantum__qis__intAsDouble(i64 %i)
-  ret double %d
+
+define void @__quantum__qis__exp__body(%Array* %.paulis, double %angle, %Array* %.qubits) {
+  %paulis = bitcast %Array* %.paulis to %struct.QirArray*
+  %qubits = bitcast %Array* %.qubits to %struct.QirArray*
+  call void @quantum__qis__exp__body(%struct.QirArray* %paulis, double %angle, %struct.QirArray* %qubits)
+  ret void
+}
+
+define void @__quantum__qis__exp__adj(%Array* %.paulis, double %angle, %Array* %.qubits) {
+  %paulis = bitcast %Array* %.paulis to %struct.QirArray*
+  %qubits = bitcast %Array* %.qubits to %struct.QirArray*
+  call void @quantum__qis__exp__adj(%struct.QirArray* %paulis, double %angle, %struct.QirArray* %qubits)
+  ret void
 }

```

Comment:
Just curiosity, feel free to ignore.
What will happen if `ret void` is removed?

Reply:
It fails to compile :)

Reply:
To be less glib with the response, I believe LLVM requires explicit ends for basic blocks, which `ret` instruction provides. And it also requires explicit type specifications for everything, including `void` so you end up with `ret void`